### ISSUE TITLE
gocv.NewMat ToImage function memory leak

### ISSUE THREAD

**Issue by snailpet** [hybridgroup/gocv#766]
gocv.NewMat ToImage function memory leak
## Description

```
func (od *OpenCVDecoder) ReadJpeg() ([]byte, error) {
	img := gocv.NewMat()
	defer func() {
		err := img.Close()
		fmt.Println("ReadJpeg err :",err)
		//runtime.GC()
		//fmt.Println("read gc")
	}()
	if ok := od.vc.Read(&img); !ok {
		return nil, nil
	}

	if img.Empty() {
		return nil, nil
	}
	_, err := img.ToImage()
	if err != nil {
		return nil, err
	}
	//i, err := ImageToJpeg(iimg)
	return nil,nil
}

```
iimg variable not release.i must use `runtime.GC()`,but it didn't work well also.i think `img.Close()` dosen't work.

after task stop. pprof report
```
(pprof) top
Showing nodes accounting for 13.85MB, 100% of 13.85MB total
      flat  flat%   sum%        cum   cum%
    7.91MB 57.13% 57.13%     7.91MB 57.13%  image.NewRGBA
    5.94MB 42.87%   100%     5.94MB 42.87%  gocv.io/x/gocv._Cfunc_GoBytes
```
```
(pprof) list ToImage
Total: 13.85MB
ROUTINE ======================== gocv.io/x/gocv.(*Mat).ToImage in /Users/x/go/pkg/mod/gocv.io/x/gocv@v0.23.0/core.go
         0    13.85MB (flat, cum)   100% of Total
         .          .    753:   }
         .          .    754:
         .          .    755:   width := m.Cols()
         .          .    756:   height := m.Rows()
         .          .    757:   step := m.Step()
         .     5.94MB    758:   data := m.ToBytes()
         .          .    759:   channels := m.Channels()
         .          .    760:
         .          .    761:   if t == MatTypeCV8UC1 {
         .          .    762:           img := image.NewGray(image.Rect(0, 0, width, height))
         .          .    763:           c := color.Gray{Y: uint8(0)}
         .          .    764:
         .          .    765:           for y := 0; y < height; y++ {
         .          .    766:                   for x := 0; x < width; x++ {
         .          .    767:                           c.Y = uint8(data[y*step+x])
         .          .    768:                           img.SetGray(x, y, c)
         .          .    769:                   }
         .          .    770:           }
         .          .    771:
         .          .    772:           return img, nil
         .          .    773:   }
         .          .    774:
         .     7.91MB    775:   img := image.NewRGBA(image.Rect(0, 0, width, height))
         .          .    776:   c := color.RGBA{
         .          .    777:           R: uint8(0),
         .          .    778:           G: uint8(0),
         .          .    779:           B: uint8(0),
         .          .    780:           A: uint8(255),

```

## Steps to Reproduce

1.
2.
3.
4.

## Your Environment

* Operating System and version: macOS
* OpenCV version used:  4.5.0
* How did you install OpenCV? brew install
* GoCV version used: 0.25.0
* Go version: 1.15.x
* Did you run the `env.sh` or `env.cmd` script before trying to `go run` or `go build`? no

**Comment by xusk** [comment]
img.close() is important

**Comment by xusk** [comment]
defer is not use in c++ opencv

**Comment by snailpet** [comment]
> img.close() is important

it dosen't work

```
	defer func() {
		err := img.Close()
		fmt.Println("ReadJpeg err :",err)
		//runtime.GC()
		//fmt.Println("read gc")
	}()
```

**Comment by xusk** [comment]
can't use defer ； you must  use img.close() every time

**Comment by xusk** [comment]
for {
		img := gocv.NewMat()
		videoCap.Read(&img)
		if img.Empty() {
			img.Close()
			break
		}
		// to do something
		img.Close()
}

and  change opencv 4.5 to opencv4.4 , Because I found opencv4.5 have memory leak

**Comment by snailpet** [comment]
> for {
> img := gocv.NewMat()
> videoCap.Read(&img)
> if img.Empty() {
> img.Close()
> break
> }
> // to do something
> img.Close()
> }
> 
> and change opencv 4.5 to opencv4.4 , Because I found opencv4.5 have memory leak

thanks.

**Comment by xusk** [comment]
> > for {
> > img := gocv.NewMat()
> > videoCap.Read(&img)
> > if img.Empty() {
> > img.Close()
> > break
> > }
> > // to do something
> > img.Close()
> > }
> > and change opencv 4.5 to opencv4.4 , Because I found opencv4.5 have memory leak
> 
> thanks. but how to install opencv 4.4? i tried `brew install opencv@4.4`, 404~

build docker image;  need cmake ffmpeg opencv  go .   brew install opencv default versions is 4.4

**Comment by snailpet** [comment]
> > > for {
> > > img := gocv.NewMat()
> > > videoCap.Read(&img)
> > > if img.Empty() {
> > > img.Close()
> > > break
> > > }
> > > // to do something
> > > img.Close()
> > > }
> > > and change opencv 4.5 to opencv4.4 , Because I found opencv4.5 have memory leak
> > 
> > 
> > thanks. but how to install opencv 4.4? i tried `brew install opencv@4.4`, 404~
> 
> build docker image; need cmake ffmpeg opencv go . brew install opencv default versions is 4.4

thanks,i tried. but it dosen't work. `ToImage` function still leak memory.

**Comment by xusk** [comment]
> > > > for {
> > > > img := gocv.NewMat()
> > > > videoCap.Read(&img)
> > > > if img.Empty() {
> > > > img.Close()
> > > > break
> > > > }
> > > > // to do something
> > > > img.Close()
> > > > }
> > > > and change opencv 4.5 to opencv4.4 , Because I found opencv4.5 have memory leak
> > > 
> > > 
> > > thanks. but how to install opencv 4.4? i tried `brew install opencv@4.4`, 404~
> > 
> > 
> > build docker image; need cmake ffmpeg opencv go . brew install opencv default versions is 4.4
> 
> thanks,i tried. but it dosen't work. `ToImage` function still leak memory.

show your code

**Comment by snailpet** [comment]
> > > > > for {
> > > > > img := gocv.NewMat()
> > > > > videoCap.Read(&img)
> > > > > if img.Empty() {
> > > > > img.Close()
> > > > > break
> > > > > }
> > > > > // to do something
> > > > > img.Close()
> > > > > }
> > > > > and change opencv 4.5 to opencv4.4 , Because I found opencv4.5 have memory leak
> > > > 
> > > > 
> > > > thanks. but how to install opencv 4.4? i tried `brew install opencv@4.4`, 404~
> > > 
> > > 
> > > build docker image; need cmake ffmpeg opencv go . brew install opencv default versions is 4.4
> > 
> > 
> > thanks,i tried. but it dosen't work. `ToImage` function still leak memory.
> 
> show your code

```
func (od *OpenCVDecoder) ReadJpeg() ([]byte, error) {
	img := gocv.NewMat()
	if ok := od.vc.Read(&img); !ok {
		_ = img.Close()
		return nil, nil
	}

	if img.Empty() {
		_ = img.Close()
		return nil, nil
	}
	iimg, err := img.ToImage()
	if err != nil {
		_ = img.Close()
		return nil, err
	}
	i, err := ImageToJpeg(iimg)
	_ = img.Close()
	return i, err
}

func ImageToJpeg(img image.Image) ([]byte, error) {
	buf := &bytes.Buffer{}
	err := jpeg.Encode(buf, img, nil)
	if err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}
```

**Comment by xusk** [comment]
> iimg

iimg close() after ImageToJpeg()

**Comment by snailpet** [comment]
iimg is a interface without close function,maybe others code make this leak. i will try  something else.

**Comment by snailpet** [comment]
I have tried many ways to solve this problem.Finally, the memory leak is located in the `Toimage` method.
```
(pprof) top
Showing nodes accounting for 27.91MB, 100% of 27.91MB total
Showing top 10 nodes out of 20
      flat  flat%   sum%        cum   cum%
   21.97MB 78.73% 78.73%    21.97MB 78.73%  bytes.makeSlice
    5.94MB 21.27%   100%     5.94MB 21.27%  gocv.io/x/gocv._Cfunc_GoBytes (inline)
         0     0%   100%    21.97MB 78.73%  bufio.(*Writer).Flush
         0     0%   100%    21.97MB 78.73%  bufio.(*Writer).WriteByte
         0     0%   100%    21.97MB 78.73%  bytes.(*Buffer).Write
         0     0%   100%    21.97MB 78.73%  bytes.(*Buffer).grow
         0     0%   100%    27.91MB   100%  git.xx.com/aisys/srv-decoder/pkg/apis/decoder.(*LiveDecoder).Output.func1
         0     0%   100%    27.91MB   100%  git.xx.com/aisys/srv-decoder/pkg/controllers/decoder.LiveDecoder
         0     0%   100%    27.91MB   100%  git.xx.com/tools/video_decode.(*OpenCVDecoder).ReadJpeg
         0     0%   100%     5.94MB 21.27%  gocv.io/x/gocv.(*Mat).ToBytes
```
Another problem is that memory leak can't be reproduced in MacOS, only in docker pod (Linux) server

**Comment by xusk** [comment]
> I have tried many ways to solve this problem.Finally, the memory leak is located in the `Toimage` method.
> 
> ```
> (pprof) top
> Showing nodes accounting for 27.91MB, 100% of 27.91MB total
> Showing top 10 nodes out of 20
>       flat  flat%   sum%        cum   cum%
>    21.97MB 78.73% 78.73%    21.97MB 78.73%  bytes.makeSlice
>     5.94MB 21.27%   100%     5.94MB 21.27%  gocv.io/x/gocv._Cfunc_GoBytes (inline)
>          0     0%   100%    21.97MB 78.73%  bufio.(*Writer).Flush
>          0     0%   100%    21.97MB 78.73%  bufio.(*Writer).WriteByte
>          0     0%   100%    21.97MB 78.73%  bytes.(*Buffer).Write
>          0     0%   100%    21.97MB 78.73%  bytes.(*Buffer).grow
>          0     0%   100%    27.91MB   100%  git.xx.com/aisys/srv-decoder/pkg/apis/decoder.(*LiveDecoder).Output.func1
>          0     0%   100%    27.91MB   100%  git.xx.com/aisys/srv-decoder/pkg/controllers/decoder.LiveDecoder
>          0     0%   100%    27.91MB   100%  git.xx.com/tools/video_decode.(*OpenCVDecoder).ReadJpeg
>          0     0%   100%     5.94MB 21.27%  gocv.io/x/gocv.(*Mat).ToBytes
> ```
> 
> Another problem is that memory leak can't be reproduced in MacOS, only in docker pod (Linux) server

mac opencv 4.4 ?  the docker is 4.5.0 ?

**Comment by snailpet** [comment]
> > I have tried many ways to solve this problem.Finally, the memory leak is located in the `Toimage` method.
> > ```
> > (pprof) top
> > Showing nodes accounting for 27.91MB, 100% of 27.91MB total
> > Showing top 10 nodes out of 20
> >       flat  flat%   sum%        cum   cum%
> >    21.97MB 78.73% 78.73%    21.97MB 78.73%  bytes.makeSlice
> >     5.94MB 21.27%   100%     5.94MB 21.27%  gocv.io/x/gocv._Cfunc_GoBytes (inline)
> >          0     0%   100%    21.97MB 78.73%  bufio.(*Writer).Flush
> >          0     0%   100%    21.97MB 78.73%  bufio.(*Writer).WriteByte
> >          0     0%   100%    21.97MB 78.73%  bytes.(*Buffer).Write
> >          0     0%   100%    21.97MB 78.73%  bytes.(*Buffer).grow
> >          0     0%   100%    27.91MB   100%  git.xx.com/aisys/srv-decoder/pkg/apis/decoder.(*LiveDecoder).Output.func1
> >          0     0%   100%    27.91MB   100%  git.xx.com/aisys/srv-decoder/pkg/controllers/decoder.LiveDecoder
> >          0     0%   100%    27.91MB   100%  git.xx.com/tools/video_decode.(*OpenCVDecoder).ReadJpeg
> >          0     0%   100%     5.94MB 21.27%  gocv.io/x/gocv.(*Mat).ToBytes
> > ```
> > 
> > 
> > Another problem is that memory leak can't be reproduced in MacOS, only in docker pod (Linux) server
> 
> mac opencv 4.4 ? the docker is 4.5.0 ?

mac 4.5.0 ,docker 4.5.0 .

**Comment by dongmu101** [comment]
mat, err := gocv.ImageToMatRGB(img)  
matClone := mat.Clone()
twoMat:=gocv.NewMat()
mat.Close()
matClone.Close()
twoMat.Close()


both version 4.5.0 and 4.4.0  leak memory

go version 1.15.5  added GODEBUG "madvdontneed=1"

**Comment by deadprogram** [comment]
Please see the latest `dev` branch, which should fix the memory leak along with some other optimizations.

**Comment by deadprogram** [comment]
This was released with v0.26.0 so now closing. Thank you everyone!
